pkg/log/locallogger/locallogger.go

```diff
@@ -1,24 +1,124 @@
+var excludedHashFields = map[string]bool{
+	"ts":     true, // go-kit timestamp
+	"time":   true, // slog timestamp
+	"caller": true, // go-kit caller info
+	"source": true, // slog source info
+}
+
+// logEntry tracks information about seen log messages for deduplication
+type logEntry struct {
+	firstSeen  time.Time
+	lastSeen   time.Time
+	count      int
+	lastLogged time.Time
+}
+
+// dedupWriter wraps the actual writer and deduplicates JSON log lines
+type dedupWriter struct {
+	localLogger  *localLogger
+	actualWriter io.Writer
+}
+
+// Write implements io.Writer and deduplicates JSON log lines before writing
+func (w *dedupWriter) Write(p []byte) (n int, err error) {
+	// Handle the case where we receive multiple log lines in one write
+	lines := strings.Split(strings.TrimSpace(string(p)), "\n")
+
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+
+		// Try to parse as JSON to extract fields for deduplication
+		var logData map[string]interface{}
+		if err := json.Unmarshal([]byte(line), &logData); err != nil {
+			// If it's not valid JSON, write it directly without deduplication
+			if _, writeErr := w.actualWriter.Write([]byte(line + "\n")); writeErr != nil {
+				return 0, writeErr
+			}
+			continue
+		}
+
+		// Create a content hash excluding timestamp and caller fields
+		hash := w.hashLogData(logData)
+
+		// Check if we should skip this duplicate
+		if skip, duplicateCount := w.localLogger.shouldSkipDuplicate(hash); skip {
+			continue // Skip this duplicate
+		} else if duplicateCount > 1 {
+			// Add duplicate count to the log data
+			logData["duplicate_count"] = duplicateCount
+			// Re-marshal with duplicate count
+			if updatedBytes, marshalErr := json.Marshal(logData); marshalErr == nil {
+				line = string(updatedBytes)
+			}
+		}
+
+		// Write the (possibly modified) log line
+		if _, writeErr := w.actualWriter.Write([]byte(line + "\n")); writeErr != nil {
+			return 0, writeErr
+		}
+	}
+
+	// Return the number of bytes we were asked to write (even if we skipped some)
+	return len(p), nil
+}
+
+// hashLogData creates a hash of the log data excluding timestamp and caller fields
+func (w *dedupWriter) hashLogData(logData map[string]interface{}) string {
+	// Convert map to key-value pairs for the shared hash function
+	var keyvals []interface{}
+	for key, value := range logData {
+		keyvals = append(keyvals, key, value)
+	}
+	return hashKeyValuePairs(keyvals...)
+}
+
 type localLogger struct {
 	logger log.Logger
 	writer io.Writer
 	lj     *lumberjack.Logger
+
+	// Deduplication fields
+	dedupCache   map[string]*logEntry
+	dedupMutex   sync.RWMutex
+	cacheExpiry  time.Duration
+	maxCacheSize int
+	lastCleanup  time.Time
+
+	// Deduplicating writer for slog and other direct writes
+	dedupWriter *dedupWriter
 }
 
-func NewKitLogger(logFilePath string) localLogger {
+func NewKitLogger(logFilePath string) *localLogger {
```

Comment:
I don't remember exactly how all the slog handlers work. Does this dedup logic living here mean that we won't be deduplicating the log entries that we ship to the cloud?